### ISSUE TITLE
Feature request: balance filesystem

### ISSUE THREAD

**Issue by stintel** [koverstreet/bcachefs#242]
When adding new disks to an old filesystem, the old disks might be close to full. It would be nice to have a way to balance the existing data equally over all disks.

Example `bcachefs fs usage` output:

```
$ bcachefs fs usage -h /srv/
Filesystem bcfb7dc1-6830-43aa-9d93-63fe7c0bbe00:
Size:                      42.9T
Used:                      25.7T
Online reserved:               0

Data type       Required/total  Devices
btree:          1/2             [sdc sdf]                        512.0K
btree:          1/1             [sda4]                           2.0M
btree:          1/2             [sda4 sdh4]                      43.0G
btree:          1/1             [sdh4]                           1.5M
btree:          1/2             [sdd sdg]                        573.0M
user:           1/1             [sdg]                            531.0G
user:           1/2             [sdg sdf]                        147.2G
user:           1/2             [sdd sdg]                        18.0T
user:           1/1             [sdf]                            1.6T
user:           1/2             [sdd sdf]                        151.1G
user:           1/1             [sdd]                            531.6G
user:           1/1             [sdc]                            1.7T
user:           1/2             [sda4 sdh4]                      44.1G
user:           1/2             [sdd sdc]                        226.0G
user:           1/2             [sdg sdc]                        199.3G
user:           1/2             [sdc sdf]                        2.5T
cached:         1/1             [sdd]                            371.2G
cached:         1/1             [sda4]                           414.8G
cached:         1/1             [sdc]                            207.3G
cached:         1/1             [sdh4]                           412.3G
cached:         1/1             [sdg]                            388.1G
cached:         1/1             [sdf]                            341.2G

hdd (device 4):                                sdc              rw
                                data         buckets      fragmented
  sb:                           2.0M               3         1020.0K
  journal:                      8.0G            8192               0
  btree:                      256.0K               1          768.0K
  user:                         3.1T         3538062          233.5G
  cached:                     207.3G           28483               0
  parity:                          0               0               0
  erasure coded:                   0               0               0
  available:                    7.5T         7897966
  capacity:                    10.9T        11444224

hdd (device 2):                                sdd              rw
                                data         buckets      fragmented
  sb:                         132.0K               1          892.0K
  journal:                    512.0M             512               0
  btree:                      286.5M             908          621.5M
  user:                         9.7T        10532435          336.9G
  cached:                     371.2G          206975               0
  parity:                          0               0               0
  erasure coded:                   0               0               0
  available:                  889.0G          910368
  capacity:                    10.9T        11444224

hdd (device 5):                                sdf              rw
                                data         buckets      fragmented
  sb:                           2.0M               3         1020.0K
  journal:                      8.0G            8192               0
  btree:                      256.0K               1          768.0K
  user:                         3.0T         3380853          217.0G
  cached:                     341.2G          181251               0
  parity:                          0               0               0
  erasure coded:                   0               0               0
  available:                    7.7T         8055175
  capacity:                    10.9T        11444224

hdd (device 3):                                sdg              rw
                                data         buckets      fragmented
  sb:                         132.0K               1          892.0K
  journal:                    512.0M             512               0
  btree:                      286.5M             908          621.5M
  user:                         9.7T        10534914          355.3G
  cached:                     388.1G          207278               0
  parity:                          0               0               0
  erasure coded:                   0               0               0
  available:                  886.6G          907889
  capacity:                    10.9T        11444224

ssd (device 0):                               sda4              rw
                                data         buckets      fragmented
  sb:                         132.0K               1          892.0K
  journal:                    512.0M             512               0
  btree:                       21.5G           44530           22.0G
  user:                        22.1G           36907           14.0G
  cached:                     414.8G          744925          312.6G
  parity:                          0               0               0
  erasure coded:                   0               0               0
  available:                    1.4T         1490914
  capacity:                     1.5T         1572864

ssd (device 1):                               sdh4              rw
                                data         buckets      fragmented
  sb:                         132.0K               1          892.0K
  journal:                    512.0M             512               0
  btree:                       21.5G           44539           22.0G
  user:                        22.1G           36904           14.0G
  cached:                     412.3G          734107          304.6G
  parity:                          0               0               0
  erasure coded:                   0               0               0
  available:                    1.4T         1490908
  capacity:                     1.5T         1572864
```

**Comment by lukasaa050** [comment]
A quick workaround is to use bcachefs device evacuate. Control-c it when balanced. And then set the disk back to rw.

**Comment by christoph-heinrich** [comment]
Instead of getting all disk to be evenly filled, sometimes it might be preferable to move just enough data around that all disks are expected to reach 100% at the same time. I don't know how much that would get used, but it would reduce the amount of data moved by a lot.

**Comment by Komzpa** [comment]
Current status from the open PR queue: I do not see a general `balance` command yet, and the current reconcile work should not be described as "evenly fill all devices".

Nearby work that may matter for the pressure/workaround cases in this issue is green/CLEAN right now:

- https://github.com/koverstreet/bcachefs-tools/pull/629 makes split foreground/background target movement self-start and keep making progress.
- https://github.com/koverstreet/bcachefs-tools/pull/641 and https://github.com/koverstreet/bcachefs-tools/pull/645 avoid starving high-priority reconcile/copygc-style movement behind idle IO priority.
- https://github.com/koverstreet/bcachefs-tools/pull/631 reduces empty reconcile worker churn and exposes reconcile fanout.
- https://github.com/koverstreet/bcachefs-tools/pull/632 persists offline evacuating scan intent, which is relevant to the `device evacuate` workaround mentioned above.

Those still do not implement the policy christoph-heinrich described, where data is moved just enough that devices are expected to reach full at the same time. That looks like the actual missing "balance" policy layer.

**Comment by Komzpa** [comment]
Follow-up after checking the newer open movement/reconcile queue: still no general balance command or policy layer found.

Two newer adjacent PRs are current-head green, but neither should be described as implementing this feature request:

- koverstreet/bcachefs-tools#686 nudges internal move allocations away from currently busy/congested devices while keeping the existing free-space/target ordering primary; head `dcd27c8f0128328216c49eb35cc5219cfd4d871b` is CLEAN with the full matrix green.
- koverstreet/bcachefs-tools#679 makes `reconcile wait` show pending work that the default wait does not wait on; head `dcad504308dd99253c5d2640fa25ee6bf8671fde` is CLEAN with the full matrix green.

The missing piece for this issue is still the higher-level balance policy discussed above: move enough data that devices are expected to reach full at the same time, rather than only making existing movement/reconcile paths less surprising or less prone to bad device choices.
